### PR TITLE
Use DNS j2 for default DNS configuration

### PR DESCRIPTION
**Pull request by ganglyu**
#### Why I did it
Support default DNS configuration

##### Work item tracking
- Microsoft ADO **(number only)**: 24408420

#### How I did it
Use j2 template to generate default DNS configuration.

#### How to verify it
Run sonic-config-engine unit test.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: src/sonic-config-engine/minigraph.py [hunk 1]
```diff
@@ -1999,6 +2000,28 @@ def parse_xml(filename, platform=None, port_config_file=None, asic_name=None, hw
...
+    # Set default DNS nameserver from dns.j2
+    results['DNS_NAMESERVER'] = {}
+    if os.environ.get("CFGGEN_UNIT_TESTING", "0") == "2":
+        dns_conf = os.path.join(os.path.dirname(__file__), "tests/", "dns.j2")
+    else:
+        dns_conf = "/usr/local/sonic/dns/dns.j2"
+    if os.path.isfile(dns_conf):
+        text = ""
+        with open(dns_conf) as template_file:
+            environment = jinja2.Environment(trim_blocks=True) # nosemgrep
```
Comment (qiluo-msft) [inline src/sonic-config-engine/minigraph.py:2012]: >nosemgrep [](http://example.com/codeflow?start=65&length=9)

In future PR, please add more code comment to explain what was the semgrep rule triggered, and why it is suppressed.

File: src/sonic-config-engine/minigraph.py [hunk 2]
```diff
@@ -1999,6 +2000,28 @@ def parse_xml(filename, platform=None, port_config_file=None, asic_name=None, hw
...
+            dns_template = environment.from_string(template_file.read())
+            text = dns_template.render(results)
+        try:
+            dns_res = json.loads(text)
+        except ValueError as e:
+            print("Warning: fail to load dns configuration, {}".format(e), file=sys.stderr)
+        else:
+            dns_nameservers = dns_res.get('DNS_NAMESERVER', {})
+            for k in dns_nameservers.keys():
+                if sys.version_info.major == 2:
```
Comment (qiluo-msft) [inline src/sonic-config-engine/minigraph.py:2022]: >version_info [](http://example.com/codeflow?start=23&length=12)

Possible to reuse UNICODE_TYPE or similar?

File: src/sonic-config-engine/minigraph.py [hunk 3]
```diff
@@ -1999,6 +2000,28 @@ def parse_xml(filename, platform=None, port_config_file=None, asic_name=None, hw
...
+    if os.path.isfile(dns_conf):
+        text = ""
+        with open(dns_conf) as template_file:
+            environment = jinja2.Environment(trim_blocks=True) # nosemgrep
+            dns_template = environment.from_string(template_file.read())
+            text = dns_template.render(results)
+        try:
+            dns_res = json.loads(text)
+        except ValueError as e:
+            print("Warning: fail to load dns configuration, {}".format(e), file=sys.stderr)
```
Comment (qiluo-msft) [inline src/sonic-config-engine/minigraph.py:2018]: >Warning [](http://example.com/codeflow?start=19&length=7)

Is it a warning or error? should we fail the whole process?

File: src/sonic-config-engine/minigraph.py (hunk 1 above)
Comment (ganglyu) [reply src/sonic-config-engine/minigraph.py:2013]: Updated with comment

File: src/sonic-config-engine/minigraph.py (hunk 2 above)
Comment (ganglyu) [reply src/sonic-config-engine/minigraph.py]: Updated, use str() instead.

File: src/sonic-config-engine/minigraph.py (hunk 3 above)
Comment (ganglyu) [reply src/sonic-config-engine/minigraph.py]: Updated, fail the whole process.

File: src/sonic-config-engine/setup.py [hunk 4]
```diff
@@ -85,6 +85,7 @@
     install_requires = dependencies,
     data_files = [
         ('/usr/share/sonic/templates', glob.glob('data/*')),
+        ('sonic/dns', ['data/dns.j2']),
```
Comment (yxieca) [inline src/sonic-config-engine/setup.py:88]: Why not place the template in the template folder above?

Why not using the absolute path?

Comment (ganglyu) [reply src/sonic-config-engine/setup.py:88]: For absolute path like '/usr/share/sonic/templates', the target folder is:
/usr/local/lib/python2.7/dist-packages/usr/share/sonic/templates/l2switch.j2
/usr/local/lib/python3.7/dist-packages/usr/share/sonic/templates/l2switch.j2
For relative path like 'sonic/dns', the target folder is: /usr/local/sonic/dns

Comment (ganglyu) [reply src/sonic-config-engine/setup.py:88]: And it's difficult to locate path like "/usr/local/lib/python3.7/dist-packages/usr/share/sonic/templates/"

**Review by qiluo-msft — APPROVED** [review]

**Review by yxieca — APPROVED** [review]

**Pull request merged**